src/controller_examples/zookeeper_controller/proof/safety.rs

```diff
@@ -961,415 +758,39 @@ pub open spec fn pending_req_in_flight_or_resp_in_flight_at_after_create_headles
-                        let resp_msg = handle_create_request(s.pending_req_of(key), s.kubernetes_api_state).1;
-                        assert(s_prime.message_in_flight(resp_msg));
-                    } else {
-                        if !s.message_in_flight(s.pending_req_of(key)) {
-                            assert(s_prime.message_in_flight(resp));
-                        }
-                    }
-                }
-                Step::ControllerStep(input) => {
-                    let cr_key = input.1.get_Some_0();
-                    if cr_key != key {
-                        if s.message_in_flight(s.pending_req_of(key)) {
-                            assert(s_prime.message_in_flight(s_prime.pending_req_of(key)));
-                        } else {
-                            assert(s_prime.message_in_flight(resp));
-                        }
-                    } else {
-                        assert(s_prime.message_in_flight(s_prime.pending_req_of(key)));
-                    }
-                }
-                Step::ClientStep(input) => {
-                    if s.message_in_flight(s.pending_req_of(key)) {
-                        assert(s_prime.message_in_flight(s_prime.pending_req_of(key)));
-                    } else {
-                        assert(s_prime.message_in_flight(resp));
-                    }
-                }
-                _ => {
-                    assert(invariant(s_prime));
-                }
-            }
-        }
-    }
-    strengthen_next::<ClusterState>(
-        spec,
-        next::<ZookeeperClusterView, ZookeeperReconcileState, ZookeeperReconciler>(),
-        each_resp_matches_at_most_one_pending_req::<ZookeeperClusterView, ZookeeperReconcileState>(key),
-        stronger_next
-    );
-    init_invariant::<ClusterState>(
-        spec,
-        init::<ZookeeperClusterView, ZookeeperReconcileState, ZookeeperReconciler>(),
-        stronger_next,
-        invariant
-    );
-}
-
-pub open spec fn pending_req_in_flight_or_resp_in_flight_at_after_create_stateful_set_step(
-    key: ObjectRef
-) -> StatePred<ClusterState>
-    recommends
-        key.kind.is_CustomResourceKind(),
-{
-    |s: ClusterState| {
-        at_after_create_stateful_set_step(key)(s)
-            ==> {
-                s.reconcile_state_of(key).pending_req_msg.is_Some()
-                && is_controller_create_request(s.pending_req_of(key))
-                && (s.message_in_flight(s.pending_req_of(key))
-                || exists |resp_msg: Message| {
-                    #[trigger] s.message_in_flight(resp_msg)
-                    && resp_msg_matches_req_msg(resp_msg, s.pending_req_of(key))
-                })
-            }
-    }
-}
-
-pub proof fn lemma_always_pending_req_in_flight_or_resp_in_flight_at_after_create_stateful_set_step(
-    spec: TempPred<ClusterState>, key: ObjectRef
+pub proof fn lemma_always_pending_req_in_flight_or_resp_in_flight_at_step(
```

Comment:
It seems to me we can even make it generally applicable to other controllers. Not super sure but we can try that later.

Reply:
I agree. But if you want to lift such lemma to controller runtime, then you have to add some postconditions to `reconcile_core`.